File: 未使用/github文档快照/github文档仓库/AI-LlmEngineerTechStack/1_大模型技术/2_模型部署与推理/硬件适配/mygpu_ops_demo/vllm_plugin/ops/rms_norm.py
"""
RMSNorm CustomOp Implementation

这个模块展示如何将RMSNorm算子包装成vLLM CustomOp。
包括标准RMSNorm和融合版本。
"""

import torch
from typing import Tuple, Optional

from .base import CustomOp, use_native_if_debug

# 导入路径处理（支持从项目目录导入mygpu_ops）
import sys
from pathlib import Path
_project_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(_project_root))


class MyGPURMSNorm(CustomOp):
    """
    MyGPU RMSNorm CustomOp
    
    这是vLLM CustomOp的标准实现模式：
    1. __init__: 存储配置参数
    2. forward_native: PyTorch原生实现
    3. forward_oot: 调用mygpu_ops实现
    
    使用示例：
        # 创建算子
        rms_norm = MyGPURMSNorm(hidden_size=4096, eps=1e-6)
        
        # 准备输入
        x = torch.randn(2, 128, 4096, dtype=torch.float16, device='cuda')
        weight = torch.ones(4096, dtype=torch.float16, device='cuda')
        
        # 调用（自动选择实现）
        output = rms_norm(x, weight)
    """
    
    def __init__(
        self,
        hidden_size: int,
        eps: float = 1e-6,
    ):
        """
        初始化RMSNorm
        
        Args:
            hidden_size: 隐藏层大小
            eps: 防止除零的小常数
        """
        super().__init__()
        self.hidden_size = hidden_size
        self.eps = eps
    
    def forward_native(
        self,
        x: torch.Tensor,
        weight: torch.Tensor,
    ) -> torch.Tensor:
        """
        PyTorch原生实现
        
        这是fallback实现，当mygpu_ops不可用时使用。
        
        数学公式：
            output = x / sqrt(mean(x^2) + eps) * weight
        
        Args:
            x: 输入张量, shape [..., hidden_size]
            weight: 归一化权重, shape [hidden_size]
        
        Returns:
            归一化后的张量
        """
        # 计算方差
        variance = x.pow(2).mean(dim=-1, keepdim=True)
        
        # 归一化
        x_normed = x * torch.rsqrt(variance + self.eps)
        
        # 乘以权重
        output = x_normed * weight
        
        return output
    
    @use_native_if_debug
    def forward_oot(
        self,
        x: torch.Tensor,
        weight: torch.Tensor,
    ) -> torch.Tensor:
        """
        Out-of-Tree实现 - 调用mygpu_ops
        
        这是性能关键路径！
        在实际场景中，这里调用的是高度优化的GPU kernel。
        
        Args:
            x: 输入张量
            weight: 权重张量
        
        Returns:
            归一化后的张量
        """
        # ========== 这是关键！调用自研算子库 ==========
        import mygpu_ops
        
        return mygpu_ops.rms_norm(x, weight, self.eps)
    
    def __repr__(self) -> str:
        return f"MyGPURMSNorm(hidden_size={self.hidden_size}, eps={self.eps})"


class MyGPUFusedAddRMSNorm(CustomOp):
    """
    MyGPU Fused Add + RMSNorm CustomOp
    
    融合残差加法和RMSNorm，减少显存访问次数。
    
    公式：
        hidden = x + residual
        output = RMSNorm(hidden) * weight
    
    使用示例：
        fused_norm = MyGPUFusedAddRMSNorm(hidden_size=4096, eps=1e-6)
        output, new_residual = fused_norm(x, residual, weight)
    """
    
    def __init__(
        self,
        hidden_size: int,
        eps: float = 1e-6,
    ):
        """初始化"""
        super().__init__()
        self.hidden_size = hidden_size
        self.eps = eps
    
    def forward_native(
        self,
        x: torch.Tensor,
        residual: torch.Tensor,
        weight: torch.Tensor,
    ) -> Tuple[torch.Tensor, torch.Tensor]:
        """
        PyTorch原生实现
        
        Args:
            x: 输入张量
            residual: 残差张量
            weight: 权重张量
        
        Returns:
            (output, new_residual)
        """
        # 残差加法
        hidden = x + residual
        new_residual = hidden
        
        # RMSNorm
        variance = hidden.pow(2).mean(dim=-1, keepdim=True)
        hidden_normed = hidden * torch.rsqrt(variance + self.eps)
        output = hidden_normed * weight
        
        return output, new_residual
    
    @use_native_if_debug
    def forward_oot(
        self,
        x: torch.Tensor,
        residual: torch.Tensor,
        weight: torch.Tensor,
    ) -> Tuple[torch.Tensor, torch.Tensor]:
        """
        Out-of-Tree实现 - 调用mygpu_ops融合kernel
        
        融合kernel的优势：
        - 减少显存带宽压力
        - 减少kernel launch开销
        - 性能提升20-30%
        """
        import mygpu_ops
        
        return mygpu_ops.fused_add_rms_norm(x, residual, weight, self.eps)
    
    def __repr__(self) -> str:
        return f"MyGPUFusedAddRMSNorm(hidden_size={self.hidden_size}, eps={self.eps})"


# ============================================================================
# 工厂函数
# ============================================================================

def create_rms_norm(
    hidden_size: int,
    eps: float = 1e-6,
    use_fused: bool = False,
) -> CustomOp:
    """
    创建RMSNorm算子的工厂函数
    
    Args:
        hidden_size: 隐藏层大小
        eps: 防止除零
        use_fused: 是否使用融合版本（需要配合残差使用）
    
    Returns:
        RMSNorm CustomOp实例
    """
    if use_fused:
        return MyGPUFusedAddRMSNorm(hidden_size=hidden_size, eps=eps)
    else:
        return MyGPURMSNorm(hidden_size=hidden_size, eps=eps)

File: 未使用/github文档快照/github文档仓库/AI-LlmEngineerTechStack/1_大模型技术/2_模型部署与推理/硬件适配/mygpu_ops_demo/csrc/ops/rms_norm.cpp
/**
 * RMSNorm C++ 实现
 * 
 * RMS LayerNorm 是大模型中最常用的归一化算子。
 * Llama、Qwen 等模型都使用它代替标准 LayerNorm。
 * 
 * 数学公式:
 *     RMSNorm(x) = x / sqrt(mean(x^2) + eps) * weight
 * 
 * 与标准 LayerNorm 的区别:
 *     - LayerNorm: (x - mean(x)) / sqrt(var(x) + eps) * weight + bias
 *     - RMSNorm: x / sqrt(mean(x^2) + eps) * weight
 *     - RMSNorm 省略了减均值和 bias，计算更快
 */

#include "../include/common.h"

namespace mygpu {
namespace ops {

/**
 * RMSNorm 前向传播 - CPU 实现
 * 
 * 这是 CPU 版本的实现，使用 ATen 操作。
 * 在实际的 GPU 实现中，这些操作会被替换为优化的 CUDA kernel。
 * 
 * @param input   输入张量, shape [..., hidden_size]
 * @param weight  归一化权重, shape [hidden_size]
 * @param eps     防止除零的小常数
 * @return        归一化后的张量, shape [..., hidden_size]
 */
torch::Tensor rms_norm_forward_cpu(
    torch::Tensor input,
    torch::Tensor weight,
    double eps
) {
    // ========== 1. 输入检查 ==========
    MYGPU_CHECK_INPUT(input);
    MYGPU_CHECK_INPUT(weight);
    
    TORCH_CHECK(input.dim() >= 2, 
                "input must have at least 2 dimensions, got ", input.dim());
    
    const int64_t hidden_size = input.size(-1);
    TORCH_CHECK(weight.size(0) == hidden_size,
                "weight size must match hidden_size, expected ", hidden_size,
                " got ", weight.size(0));
    
    // ========== 2. 保存原始形状并展平 ==========
    // 将 [..., hidden_size] 展平为 [num_rows, hidden_size]
    auto original_shape = input.sizes().vec();
    const int64_t num_rows = utils::get_num_rows(input);
    
    auto input_2d = input.view({num_rows, hidden_size});
    
    // ========== 3. 计算 RMSNorm ==========
    // 在实际 GPU 实现中，这三步会融合成一个 kernel
    
    // Step 1: 计算 variance = mean(x^2)
    // 对最后一个维度求平方和的均值
    auto variance = input_2d.pow(2).mean(/*dim=*/-1, /*keepdim=*/true);
    
    // Step 2: 计算 rsqrt(variance + eps)
    auto rsqrt_var = torch::rsqrt(variance + eps);
    
    // Step 3: 归一化并乘以权重
    auto output = input_2d * rsqrt_var * weight;
    
    // ========== 4. 恢复原始形状 ==========
    return output.view(original_shape);
}


/**
 * Fused Add RMSNorm 前向传播 - CPU 实现
 * 
 * 融合残差加法和 RMSNorm，减少显存访问次数。
 * 
 * 公式:
 *     hidden = input + residual
 *     output = RMSNorm(hidden) * weight
 * 
 * 融合优化原理:
 *     - 非融合: 读input + 读residual + 写sum + 读sum + 写output = 5次显存访问
 *     - 融合: 读input + 读residual + 写output + 写residual = 4次显存访问
 *     - 理论性能提升 ~20%
 * 
 * @param input    输入张量
 * @param residual 残差张量
 * @param weight   权重张量
 * @param eps      防止除零
 * @return         (output, new_residual)
 */
std::tuple<torch::Tensor, torch::Tensor> fused_add_rms_norm_forward_cpu(
    torch::Tensor input,
    torch::Tensor residual,
    torch::Tensor weight,
    double eps
) {
    // ========== 1. 输入检查 ==========
    MYGPU_CHECK_INPUT(input);
    MYGPU_CHECK_INPUT(residual);
    MYGPU_CHECK_INPUT(weight);
    
    TORCH_CHECK(input.sizes() == residual.sizes(),
                "input and residual must have the same shape");
    
    const int64_t hidden_size = input.size(-1);
    TORCH_CHECK(weight.size(0) == hidden_size,
                "weight size must match hidden_size");
    
    // ========== 2. 展平 ==========
    auto original_shape = input.sizes().vec();
    const int64_t num_rows = utils::get_num_rows(input);
    
    auto input_2d = input.view({num_rows, hidden_size});
    auto residual_2d = residual.view({num_rows, hidden_size});
    
    // ========== 3. 融合计算 ==========
    // 在实际 GPU 实现中，这些操作在一个 kernel 中完成
    
    // Step 1: 残差加法
    auto hidden = input_2d + residual_2d;
    
    // Step 2: 保存新的残差（用于下一层）
    // 注意：这里使用 clone() 确保返回独立的张量
    auto new_residual = hidden.clone();
    
    // Step 3: RMSNorm
    auto variance = hidden.pow(2).mean(/*dim=*/-1, /*keepdim=*/true);
    auto hidden_normed = hidden * torch::rsqrt(variance + eps);
    auto output = hidden_normed * weight;
    
    // ========== 4. 恢复形状 ==========
    return std::make_tuple(
        output.view(original_shape),
        new_residual.view(original_shape)
    );
}

}  // namespace ops
}  // namespace mygpu


// =============================================================================
// 导出函数（供 pybind11 绑定使用）
// =============================================================================

torch::Tensor rms_norm_forward(
    torch::Tensor input,
    torch::Tensor weight,
    double eps
) {
    // 根据设备选择实现
    // 目前只有 CPU 实现，未来可以添加 CUDA 实现
    if (input.device().is_cpu()) {
        return mygpu::ops::rms_norm_forward_cpu(input, weight, eps);
    }
    
    // 对于 CUDA 设备，暂时回退到 CPU 计算
    // 在实际实现中，这里会调用 CUDA kernel
    TORCH_WARN("CUDA implementation not available, falling back to CPU");
    auto cpu_input = input.cpu();
    auto cpu_weight = weight.cpu();
    auto cpu_output = mygpu::ops::rms_norm_forward_cpu(cpu_input, cpu_weight, eps);
    return cpu_output.to(input.device());
}

std::tuple<torch::Tensor, torch::Tensor> fused_add_rms_norm_forward(
    torch::Tensor input,
    torch::Tensor residual,
    torch::Tensor weight,
    double eps
) {
    if (input.device().is_cpu()) {
        return mygpu::ops::fused_add_rms_norm_forward_cpu(input, residual, weight, eps);
    }
    
    TORCH_WARN("CUDA implementation not available, falling back to CPU");
    auto cpu_input = input.cpu();
    auto cpu_residual = residual.cpu();
    auto cpu_weight = weight.cpu();
    auto [cpu_output, cpu_new_residual] = 
        mygpu::ops::fused_add_rms_norm_forward_cpu(cpu_input, cpu_residual, cpu_weight, eps);
    return std::make_tuple(
        cpu_output.to(input.device()),
        cpu_new_residual.to(input.device())
    );
}
